File: server/internal/config/cron.go
// internal/config/cron.go
package config

import (
	"strings"

	"github.com/robfig/cron/v3"
)

// cronParser mirrors the field spec scheduler.NewScheduler configures via
// cron.WithSeconds() -- kept in sync deliberately, so that a schedule
// accepted here (at save time) can never be rejected by the scheduler
// itself at run time, or vice versa.
var cronParser = cron.NewParser(
	cron.Second | cron.Minute | cron.Hour | cron.Dom | cron.Month | cron.Dow | cron.Descriptor,
)

// ValidateCronSchedule reports whether expr parses as a valid cron
// expression under the scheduler's field spec (seconds field included,
// e.g. "59 59 23 * * FRI"). Used to reject a bad schedule entry at
// save time instead of only failing silently when the scheduler tries to
// register it.
func ValidateCronSchedule(expr string) error {
	_, err := cronParser.Parse(expr)
	return err
}

// NormalizeCronExpression collapses runs of whitespace and lowercases a
// cron expression, so two expressions that only differ by spacing or by
// day/month name casing (e.g. "FRI" vs "fri") compare equal -- used to
// detect a duplicate schedule entry for the same miner.
func NormalizeCronExpression(expr string) string {
	return strings.ToLower(strings.Join(strings.Fields(expr), " "))
}

File: ui/src/utils/cron.ts
// src/utils/cron.ts
import { CronExpressionParser } from "cron-parser";
import cronstrue from "cronstrue";

import "cronstrue/locales/fr";

// Mirrors the field spec the server-side scheduler and config.ValidateCronSchedule
// use (seconds field included, e.g. "59 59 23 * * FRI") -- everything here
// is purely for the live preview in ScheduleEditor, the actual save is
// still validated server-side.

/** Human-readable translation of a cron expression (e.g. "At 23:59:59, only
 * on Friday"), in the given locale ("en" or "fr"), or null if it doesn't
 * parse. */
export function describeCron(expr: string, locale: string): string | null {
  const trimmed = expr.trim();
  if (!trimmed) return null;
  try {
    return cronstrue.toString(trimmed, {
      throwExceptionOnParseError: true,
      locale: locale.startsWith("fr") ? "fr" : "en",
    });
  } catch {
    return null;
  }
}

/** The next `count` dates this cron expression would fire at, or null if it
 * doesn't parse. */
export function nextCronRuns(expr: string, count = 3): Date[] | null {
  const trimmed = expr.trim();
  if (!trimmed) return null;
  try {
    const interval = CronExpressionParser.parse(trimmed);
    return Array.from({ length: count }, () => interval.next().toDate());
  } catch {
    return null;
  }
}

/** Collapses whitespace and lowercases a cron expression, so two
 * expressions that only differ by spacing or by day/month name casing
 * (e.g. "FRI" vs "fri") compare equal -- mirrors
 * config.NormalizeCronExpression server-side. Used to catch a duplicate
 * schedule entry before it's even submitted. */
export function normalizeCronExpression(expr: string): string {
  return expr.trim().split(/\s+/).join(" ").toLowerCase();
}
